File: ios/Classes/Data/Beacon.swift
//
//  Copyright (c) 2018 Loup Inc.
//  Licensed under Apache License v2.0
//
import Foundation
import CoreLocation
import SensoroBeaconKit

struct Beacon : Codable {
    let ids: [AnyCodable]
    let distance: Double
    let rssi: Int
    let txPower: Int
    let battery: Double
    let temperature: Int
    let platformCustoms: [String:AnyCodable]
    
    //  init(from beacon: CLBeacon) {
    //    self.ids = [
    //      AnyCodable(beacon.proximityUUID.uuidString),
    //      AnyCodable(beacon.major.intValue),
    //      AnyCodable(beacon.minor.intValue)
    //    ]
    //
    //    self.distance = beacon.accuracy
    //    self.rssi = beacon.rssi
    //    self.txPower = 0
    //    self.battery = 0
    //    self.temperature = 0
    //    self.platformCustoms = [
    //      "proximity": AnyCodable(Proximity(from: beacon.proximity))
    //    ]
    //  }
    
    init(from beacon: Any) {
        switch beacon {
        case let sbkBeacon as SBKBeacon:
            self.ids = [
                AnyCodable(sbkBeacon.beaconID.proximityUUID.uuidString),
                AnyCodable(sbkBeacon.beaconID.major.intValue),
                AnyCodable(sbkBeacon.beaconID.minor.intValue)
            ]
            
            self.distance = sbkBeacon.accuracy
            self.rssi = sbkBeacon.rssi
            self.txPower = 0
            self.battery = sbkBeacon.batteryLevel.doubleValue
            self.temperature = sbkBeacon.temperature.intValue
            self.platformCustoms = [
                "proximity": AnyCodable(Proximity(from: sbkBeacon.proximity)),
                "powerLevel": AnyCodable(sbkBeacon.broadcastTransmitPower)
            ]
        case let clBeacon as CLBeacon:
            self.ids = [
                AnyCodable(clBeacon.proximityUUID.uuidString),
                AnyCodable(clBeacon.major.intValue),
                AnyCodable(clBeacon.minor.intValue)
            ]
            
            self.distance = clBeacon.accuracy
            self.rssi = clBeacon.rssi
            self.txPower = 0
            self.battery = -1.0
            self.temperature = 0
            self.platformCustoms = [
                "proximity": AnyCodable(Proximity(from: clBeacon.proximity))
            ]
        default:
            self.ids = [];
            self.distance = 0
            self.rssi = 0
            self.txPower = 0
            self.battery = -1.0
            self.temperature = 0
            self.platformCustoms = [
                "proximity": "unknown"
            ]
        }
    }
}

File: lib/data/beacon.dart
//  Copyright (c) 2018 Loup Inc.
//  Licensed under Apache License v2.0

part of beacons;

class Beacon {
  Beacon._(
    this.ids,
    this.distance,
    this.rssi,
    this.mRssi,
    this.battery,
    this.temperature,
    this.powerLevel,
    this._platformCustoms,
  );

  final List<dynamic> ids;
  final double distance;
  final int rssi;
  final int mRssi;
  final double battery;
  final int temperature;
  final int powerLevel;
  final Map<String, dynamic> _platformCustoms;
}

class BeaconIBeacon {
  BeaconIBeacon.from(Beacon beacon)
      : proximityUUID = beacon.ids[0],
        major = beacon.ids.length > 1 ? beacon.ids[1] : null,
        minor = beacon.ids.length > 2 ? beacon.ids[2] : null,
        accuracy = beacon.distance,
        rssi = beacon.rssi,
        mRssi = beacon.mRssi,
        battery = beacon.battery != null ? beacon.battery : -1.0,
        temperature = beacon.temperature != null ? beacon.temperature : -999,
        proximity =
            _JsonCodec.proximityFromJson(beacon._platformCustoms['proximity']),
        powerLevel = beacon.powerLevel != null ? beacon.powerLevel : -1,
        createdAt = DateTime.now();

  final String proximityUUID;
  final int major;
  final int minor;
  final double accuracy;
  final int rssi;
  final int mRssi;
  final double battery;
  final int temperature;
  final BeaconProximity proximity;
  final int powerLevel;
  DateTime createdAt;
}

enum BeaconProximity {
  unknown,
  immediate,
  near,
  far,
}
